project-koku/koku
# 1259 Azure data removal

reference #1259 

This PR implements Azure data removal from the database when the Azure provider is deleted.

Testing:
1. Add Azure provider.
2. Ingest Azure data.
3. Delete Azure provider.
4. Check Azure database tables and see that Azure data was removed.

Results (starting with empty db):
1/2. Add and ingest:

```
postgres=# select * from acct10001.reporting_azurecostentrybill LIMIT 1;
-[ RECORD 1 ]------------------+------------------------------
id                             | 1
billing_period_start           | 2019-09-01 00:00:00+00
billing_period_end             | 2019-09-30 00:00:00+00
summary_data_creation_datetime | 2019-10-18 18:07:28.744877+00
summary_data_updated_datetime  | 2019-10-18 18:07:28.744893+00
finalized_datetime             | 
derived_cost_datetime          | 2019-10-18 18:07:28.80312+00
provider_id                    | 1

postgres=# select * from acct10001.reporting_azurecostentrylineitem_daily LIMIT 1;
-[ RECORD 1 ]---------+-------------------------------------
id                    | 1
subscription_guid     | uuid
tags                  | {"project": "p1"}
usage_date_time       | 2019-09-14 00:00:00+00
usage_quantity        | 0.000100000
pretax_cost           | 0.000000000
offer_id              | 
cost_entry_bill_id    | 1
cost_entry_product_id | 1
meter_id              | 1

postgres=# select * from acct10001.reporting_azurecostentrylineitem_daily_summary LIMIT 1;
-[ RECORD 1 ]------+---------------------------------------------------------------------------------------------------------------------------------------
id                 | 1067
subscription_guid  | uuid
instance_type      | 
service_name       | Storage
resource_location  | US East 2
tags               | {"project": "p1"}
usage_start        | 2019-10-01 00:00:00+00
usage_quantity     | 0.000100000
pretax_cost        | 0.000000000
offer_id           | 
cost_entry_bill_id | 2
meter_id           | 1
currency           | USD
instance_count     | 1
instance_ids       | {/subscriptions/uuid/resourceGroups/costmgmt/providers/Microsoft.Storage/storageAccounts/costmgmtdiag}
usage_end          | 2019-10-01 00:00:00+00
markup_cost        | 0.000000000
unit_of_measure    | 
```
3. Delete provider:
```
koku_server       | [2019-10-18 18:13:54,856] INFO: Calling masu to delete report data for provider 1
masu_server       | [2019-10-18 18:13:54,868] INFO: Calling remove_expired_data async task.
koku_worker       | [2019-10-18 18:13:54,886: INFO/ForkPoolWorker-1] masu.processor.tasks.remove_expired_data[58fb7b7b-e3d4-47db-b98a-3da19b4d6f12]: remove_expired_data called with args:
koku_worker       |  schema_name: acct10001,
koku_worker       |  provider: AZURE,
koku_worker       |  simulate: False,
koku_worker       |  provider_id: 1
koku_worker       | [2019-10-18 18:13:54,887: INFO/ForkPoolWorker-1] masu.processor.tasks.remove_expired_data[58fb7b7b-e3d4-47db-b98a-3da19b4d6f12]: Remove expired data:
koku_worker       |  schema_name: acct10001
koku_worker       |  provider: AZURE
koku_worker       |  simulate: False
koku_worker       | [2019-10-18 18:13:54,887] INFO: Current date time is 2019-10-18 18:13:54.887147+00:00
koku_worker       | [2019-10-18 18:13:54,889] INFO: Report data expiration is 2019-07-01 00:00:00+00:00 for a 3 month retention policy
koku_worker       | [2019-10-18 18:13:54,889] INFO: Calling purge_expired_report_data
koku_server       | [2019-10-18 18:13:54,898] INFO: Response: {'Report Data Task ID': '58fb7b7b-e3d4-47db-b98a-3da19b4d6f12'}
masu_server       | [2019-10-18 18:13:54,896] INFO: "DELETE /api/cost-management/v1/report_data/?schema=acct10001&provider=AZURE&provider_id=1 HTTP/1.1" 200 62
koku_worker       | [2019-10-18 18:13:54,906] INFO: Removing (1073, {'reporting.AzureCostEntryLineItemDaily': 1073}) cost entry line items for bill id 1
koku_server       | [2019-10-18 18:13:54,908] INFO: Provider: Test Azure Source removed by user_dev
koku_worker       | [2019-10-18 18:13:54,911] INFO: Removing (1066, {'reporting.AzureCostEntryLineItemDailySummary': 1066}) cost entry summary items for bill id 1
koku_worker       | [2019-10-18 18:13:54,911] INFO: Report data removed for Account Payer ID: 1 with billing period: 2019-09-01 00:00:00+00:00
koku_worker       | [2019-10-18 18:13:54,914] INFO: Removing (1232, {'reporting.AzureCostEntryLineItemDaily': 1232}) cost entry line items for bill id 2
koku_worker       | [2019-10-18 18:13:54,917] INFO: Removing (1216, {'reporting.AzureCostEntryLineItemDailySummary': 1216}) cost entry summary items for bill id 2
koku_worker       | [2019-10-18 18:13:54,918] INFO: Report data removed for Account Payer ID: 1 with billing period: 2019-10-01 00:00:00+00:00
koku_server       | [2019-10-18 18:13:54,918] INFO: "DELETE /api/cost-management/v1/providers/7c4cf36e-5458-4493-8e88-c4d7b99d05e0/ HTTP/1.1" 204 0
koku_worker       | [2019-10-18 18:13:54,934: INFO/ForkPoolWorker-1] masu.processor.tasks.remove_expired_data[58fb7b7b-e3d4-47db-b98a-3da19b4d6f12]: Removed Data:
koku_worker       |  [{'provider_id': 1, 'billing_period_start': '2019-09-01 00:00:00+00:00'}, {'provider_id': 1, 'billing_period_start': '2019-10-01 00:00:00+00:00'}]
```

4. check db:
```
postgres=# select * from acct10001.reporting_azurecostentrybill LIMIT 1;
(0 rows)

postgres=# select * from acct10001.reporting_azurecostentrylineitem_daily LIMIT 1;
(0 rows)

postgres=# select * from acct10001.reporting_azurecostentrylineitem_daily_summary LIMIT 1;
(0 rows)
```
